cmd/openstack-cloud-controller-manager/main.go

```diff
@@ -20,240 +20,56 @@ limitations under the License.
-	"k8s.io/kubernetes/cmd/cloud-controller-manager/app/options"
-	cloudcontrollers "k8s.io/kubernetes/pkg/controller/cloud"
-	routecontroller "k8s.io/kubernetes/pkg/controller/route"
-	servicecontroller "k8s.io/kubernetes/pkg/controller/service"
 	_ "k8s.io/kubernetes/pkg/features" // add the kubernetes feature gates
-	utilflag "k8s.io/kubernetes/pkg/util/flag"
-
-	"github.com/spf13/cobra"
-	"github.com/spf13/pflag"
 )
 
 func init() {
 	mux := http.NewServeMux()
 	healthz.InstallHandler(mux)
 }
 
-var (
-	versionFlag bool
-)
-
 func main() {
 	rand.Seed(time.Now().UTC().UnixNano())
 
-	goflag.CommandLine.Parse([]string{})
-	s, err := options.NewCloudControllerManagerOptions()
-	if err != nil {
-		klog.Fatalf("unable to initialize command options: %v", err)
-	}
-
-	command := &cobra.Command{
-		Use: "openstack-cloud-controller-manager",
-		Long: `The Cloud controller manager is a daemon that embeds
-the cloud specific control loops shipped with Kubernetes.`,
-		PersistentPreRun: func(cmd *cobra.Command, args []string) {
-			// Glog requires this otherwise it complains.
-			goflag.CommandLine.Parse(nil)
-
-			// This is a temporary hack to enable proper logging until upstream dependencies
-			// are migrated to fully utilize klog instead of glog.
-			klogFlags := goflag.NewFlagSet("klog", goflag.ExitOnError)
-			klog.InitFlags(klogFlags)
-
-			// Sync the glog and klog flags.
-			cmd.Flags().VisitAll(func(f1 *pflag.Flag) {
-				f2 := klogFlags.Lookup(f1.Name)
-				if f2 != nil {
-					value := f1.Value.String()
-					f2.Value.Set(value)
-				}
-			})
-		},
-		Run: func(cmd *cobra.Command, args []string) {
-			if versionFlag {
-				version.PrintVersionAndExit()
-			}
-
-			utilflag.PrintFlags(cmd.Flags())
-
-			c, err := s.Config(KnownControllers(), ControllersDisabledByDefault.List())
-			if err != nil {
-				fmt.Fprintf(os.Stderr, "%v\n", err)
-				os.Exit(1)
-			}
-
-			if err := app.Run(c.Complete(), wait.NeverStop); err != nil {
-				fmt.Fprintf(os.Stderr, "%v\n", err)
-				os.Exit(1)
-			}
-		},
-	}
-
-	fs := command.Flags()
-	namedFlagSets := s.Flags(KnownControllers(), ControllersDisabledByDefault.List())
-	for _, f := range namedFlagSets.FlagSets {
-		fs.AddFlagSet(f)
-	}
-
-	fs.BoolVar(&versionFlag, "version", false, "Print version and exit")
-
-	openstack.AddExtraFlags(pflag.CommandLine)
+	command := app.NewCloudControllerManagerCommand()
 
 	// TODO: once we switch everything over to Cobra commands, we can go back to calling
 	// utilflag.InitFlags() (by removing its pflag.Parse() call). For now, we have to set the
 	// normalize func and add the go flag set by hand.
-	pflag.CommandLine.SetNormalizeFunc(flag.WordSepNormalizeFunc)
-	pflag.CommandLine.AddGoFlagSet(goflag.CommandLine)
 	// utilflag.InitFlags()
 	logs.InitLogs()
 	defer logs.FlushLogs()
 
-	klog.V(1).Infof("openstack-cloud-controller-manager version: %s", version.Version)
-
-	s.KubeCloudShared.CloudProvider.Name = openstack.ProviderName
```

Comment:
I'm a bit unsure here, but i think this is set while using the `--cloud-provider` flag, which in this commit is done by parsing all command line flags and if `cloud-provider` is found i set that to `openstack` using the `openstack.ProviderName` variable.